autogluon/utils/tabular/metrics/classification_metrics.py

```diff
@@ -17,15 +17,19 @@ def balanced_accuracy(solution, prediction):
         pass
 
     elif y_type == 'multiclass':
-        # Need to create a multiclass solution and a multiclass predictions
-        max_class = int(np.max((np.max(solution), np.max(prediction))))
-        solution_binary = np.zeros((len(solution), max_class + 1))
-        prediction_binary = np.zeros((len(prediction), max_class + 1))
-        for i in range(len(solution)):
-            solution_binary[i, int(solution[i])] = 1
-            prediction_binary[i, int(prediction[i])] = 1
-        solution = solution_binary
-        prediction = prediction_binary
+        n = len(solution)
+        unique_sol, encoded_sol = np.unique(solution, return_inverse=True)
+        unique_pred, encoded_pred = np.unique(prediction, return_inverse=True)
+        classes = np.unique(np.concatenate((unique_sol, unique_pred)))
+        map_sol = np.array([np.where(classes==c)[0][0] for c in unique_sol])
+        map_pred = np.array([np.where(classes==c)[0][0] for c in unique_pred])
+        # one hot encoding
+        sol_ohe = np.zeros((n, len(classes)))
```

Comment:
A bit concerned with OHE, if num classes is very large (say 1000), this will become huge. Is there any way to avoid OHE? I think a performance benchmark example with runtimes of sklearn vs ag on this metric could be helpful (with artificially very large class count inputs)

Reply:
sklearn source code for reference:

```
    C = confusion_matrix(y_true, y_pred, sample_weight=sample_weight)
    with np.errstate(divide='ignore', invalid='ignore'):
        per_class = np.diag(C) / C.sum(axis=1)
    if np.any(np.isnan(per_class)):
        warnings.warn('y_pred contains classes not in y_true')
        per_class = per_class[~np.isnan(per_class)]
    score = np.mean(per_class)
    if adjusted:
        n_classes = len(per_class)
        chance = 1 / n_classes
        score -= chance
        score /= 1 - chance
    return score
```

Reply:
Furthermore, does it make sense to not error if a class is missing? Are we sure sklearn's implementation is not correct in this instance?

Reply:
1. we could add logic to compute a confusion matrix, this would be useful because we could compute a whole bunch of derived metrics out of it, note that the confusion matrix in multi class is C x C where C is the number of classes so if you have 1000 classes, you still have to pay that price. Note also that the previous code, afaik, was basically OHE based as well so this is just a fix, but we could open another issue and fix this proper. I would be in favour of adding logic to compute the confusion matrix in general and then have methods on that
2. in terms of sklearn's implementation, this is a somewhat tricky topic, the behaviour now is what I believe the intended behaviour was before I made the change so I kept it that way. IMO it makes sense to have this (slightly) more general thing: you could train on 100 classes but then make a prediction on something where you only predict 50 classes and the ground truth has 45. This would not happen in a situation where you have scores for every class seen at training though. 

So basically I would suggest the following steps:

1. merge this for now (possibly with minor modification) as it's the past behaviour but actually correct in the general case (I think)
2. open another issue discussing confusion matrix + derived metrics

Reply:
Good points and I'm aligned on your next-steps. Thanks for the contribution!